tests/test_cache.py

```diff
@@ -0,0 +1,105 @@
+import httpretty
+import redis
+import shutil
+import unittest
+import requests
+from cachecontrol.caches.file_cache import FileCache
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+
+file_cache_directory = ".testcache"
+
+
+def create_redis_connection_pool(
+    redis_port="6379", redis_host="localhost", redis_db=0
+):
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
+
+
+class TestCachedSession(unittest.TestCase):
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(
+            fallback_cache_duration=10,
+            file_cache_directory=file_cache_directory,
+        )
+        resp = session.get("https://httpbin.org/ip")
+
+        for h in resp.headers:
+            if h == "cache-control":
+                print(resp.headers[h])
+
+        self.assertIn(
+            "Automatically cached for 0:00:10", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_default_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertIn(
+            "Automatically cached for 0:00:05", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_cache_control_max_age_overwrites_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+            adding_headers={"Cache-Control": "max-age=30"},
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("max-age=30", resp.headers["cache-control"])
```

Comment:
Again, checking that the header exists is a no-brainer - it proves nothing. What we need to check is that the response is actually cached. Similarly to as described above.